internal/pkg/vm/internal/vmcontext/invocation_context.go

```diff
@@ -0,0 +1,311 @@
+package vmcontext
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/filecoin-project/go-filecoin/internal/pkg/types"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/abi"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/actor"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/actor/builtin/initactor"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/address"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/internal/exitcode"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/internal/gas"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/internal/gascost"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/internal/message"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/internal/runtime"
+	"github.com/ipfs/go-cid"
+)
+
+type invocationContext struct {
+	rt                *VM
+	msg               internalMessage
+	fromActor         *actor.Actor
+	gasTank           *gas.Tracker
+	isCallerValidated bool
+	allowSideEffects  bool
+	toActor           *actor.Actor
+	stateHandle       internalActorStateHandle
+}
+
+type internalActorStateHandle interface {
+	runtime.ActorStateHandle
+	Validate()
+}
+
+func newInvocationContext(rt *VM, msg internalMessage, fromActor *actor.Actor, gasTank *gas.Tracker) invocationContext {
+	// Note: the toActor and stateHandle are loaded during the `invoke()`
+	return invocationContext{
+		rt:                rt,
+		msg:               msg,
+		fromActor:         fromActor,
+		gasTank:           gasTank,
+		isCallerValidated: false,
+		allowSideEffects:  false,
+	}
+}
+
+type stateHandleContext invocationContext
+
+func (ctx *stateHandleContext) AllowSideEffects(allow bool) {
+	ctx.allowSideEffects = allow
+}
+
+func (ctx *stateHandleContext) LegacyStorage() runtime.LegacyStorage {
+	return ctx.rt.LegacyStorage()
+}
+
+func (ctx *invocationContext) invoke() interface{} {
+	// pre-dispatch
+	// 1. charge gas for message invocation
+	// 2. load target actor
+	// 3. transfer optional funds
+	// 4. short-circuit _Send_ method
+	// 5. load target actor code
+	// 6. create target state handle
+
+	// Dragons: assert from address is an id address.
+	// TODO: move the resolve target outisde, assert the to address is an id address.
+
+	// 1. charge gas for msg
+	ctx.gasTank.Charge(gascost.OnMethodInvocation(&ctx.msg))
+
+	// 2. load target actor
+	// Note: we replace the "to" address with the normalized version
+	ctx.toActor, ctx.msg.to = ctx.resolveTarget(ctx.msg.to)
+
+	// 3. transfer funds carried by the msg
+	ctx.rt.transfer(ctx.msg.from, ctx.msg.to, ctx.msg.value)
+
+	// 4. if we are just sending funds, there is nothing else to do.
+	if ctx.msg.method == types.SendMethodID {
+		return message.Ok().WithGas(ctx.gasTank.GasConsumed())
+	}
+
+	// 5. load target actor code
+	// TODO: use chain height based protocol version here (#3360)
+	actorImpl := ctx.rt.getActorImpl(ctx.toActor.Code)
```

Comment:
Perhaps pass the height as a parameter here so we can hook it up later.

Reply:
it already is, that method is a helper in the context, it internally passes the height.
```go
func (vm *VM) getActorImpl(code cid.Cid) dispatch.ExecutableActor {
	actorImpl, err := vm.actorImpls.GetActorImpl(code, vm.currentEpoch)
	if err != nil {
		exitcode.AbortWithCode(exitcode.ActorCodeNotFound)
	}
	return actorImpl
}
```

Reply:
I think the comment is a copy/pasta from the legacy code. I'll remove it.